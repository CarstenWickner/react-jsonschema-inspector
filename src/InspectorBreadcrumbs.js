import PropTypes from "prop-types";
import React from "react";
import classNames from "classnames";

import JsonSchema from "./JsonSchema";
import { isDefined } from "./utils";

const InspectorBreadcrumbs = ({
    columnData, prefix, separator, arrayItemAccessor, preventNavigation
}) => (
    <div className="jsonschema-inspector-breadcrumbs">
        <span className="jsonschema-inspector-breadcrumbs-icon" />
        {columnData.map(({
            items, selectedItem, trailingSelection, onSelect
        }, index) => {
            if (!selectedItem) {
                // no selection: nothing to show in the breadcrumbs
                return null;
            }
            // if this is not the last selection, we need to add a (dummy) array item selector for each nested array
            let suffix = "";
            if (!trailingSelection) {
                let itemSchema = items[selectedItem].getTypeOfArrayItems();
                while (isDefined(itemSchema)) {
                    suffix += arrayItemAccessor;
                    itemSchema = itemSchema.getTypeOfArrayItems();
                }
            }
            const className = classNames({
                "jsonschema-inspector-breadcrumbs-item": true,
                "has-nested-items": index < (columnData.length - 1)
            });
            return (
                <span
                    // eslint-disable-next-line react/no-array-index-key
                    key={index}
                    className={className}
                    onDoubleClick={preventNavigation ? undefined : event => onSelect(event, selectedItem)}
                >
                    {(index === 0 ? prefix : separator) + selectedItem + suffix}
                </span>
            );
        })}
    </div>
);

InspectorBreadcrumbs.propTypes = {
    columnData: PropTypes.arrayOf(PropTypes.shape({
        items: PropTypes.objectOf(PropTypes.instanceOf(JsonSchema)).isRequired,
        selectedItem: PropTypes.string,
        trailingSelection: PropTypes.bool,
        onSelect: PropTypes.func.isRequired
    })).isRequired,
    prefix: PropTypes.string,
    separator: PropTypes.string,
    arrayItemAccessor: PropTypes.string,
    preventNavigation: PropTypes.bool
};

InspectorBreadcrumbs.defaultProps = {
    prefix: "",
    separator: ".",
    arrayItemAccessor: "[0]",
    preventNavigation: false
};

export default InspectorBreadcrumbs;
