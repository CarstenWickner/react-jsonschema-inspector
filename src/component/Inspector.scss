.jsonschema-inspector {
    display: flex;
    flex-direction: column;
    font-size: 13px;
    font-family: Arial, Helvetica, sans-serif;

    $columnWidth: 15em;
    $detailsWidth: 20em;
    $selectionColor: #dcdcdc;
    $trailingSelectionColor: #276bd2;
    $backgroundColor: #fff;
    $innerBorderColor: #e0e0e0;
    $outerBorderColor: #cbcbcb;
    $innerBorder: 1px solid $innerBorderColor;
    $outerBorder: 1px solid $outerBorderColor;

    $verticalSpacingOutside: 0.25em;
    $verticalItemSpacingInside: 0.25em;
    $leadItemSpacingInside: 0.75em;
    $spacingInside: 0.5em;
    $borderRadius: 0.25em;

    &-header,
    &-body,
    &-footer {
        display: flex;
        background-color: $backgroundColor;
        border: $outerBorder;
        border-radius: $borderRadius;
    }
    &-body {
        height: 25em;
    }
    &-colview {
        flex-grow: 1;
        display: flex;
        overflow-x: auto;
        margin-right: -1px;
    }
    &-column {
        &,
        &-placeholder {
            display: flex;
            flex-direction: column;
            width: $columnWidth;
            min-width: $columnWidth;
            border-right: $innerBorder;
            overflow-y: auto;
        }
        &-placeholder {
            opacity: 0;
        }
        .optional-group-title {
            display: block;
            padding: $verticalItemSpacingInside $spacingInside $verticalItemSpacingInside $leadItemSpacingInside;
            font-style: italic;
        }
        &.optional-groups ul {
            margin: 0;
            padding: 0;
            list-style-type: none;

            ul {
                background-color: $innerBorderColor;
            }
            ul,
            .jsonschema-inspector-item {
                padding-left: $leadItemSpacingInside;
            }
            ul .optional-group-title,
            ul .jsonschema-inspector-item {
                background-color: $backgroundColor;
            }
            .optional-group-title {
                padding-left: (2 * $leadItemSpacingInside);
            }
        }
    }
    &-item {
        flex-shrink: 0;
        text-align: start;
        border: none;
        background: inherit;
        color: inherit;
        font: inherit;
        padding: 0;
        margin: 0;
        width: 100%;

        &.selected {
            background-color: $selectionColor;
        }
        &.matching-filter {
            font-weight: bold;
        }
        &.not-matching-filter {
            opacity: 0.5;
        }
        &-content {
            display: flex;
            flex-direction: row;
        }
        &-name {
            display: block;
            flex-grow: 1;
            padding: $verticalItemSpacingInside $spacingInside $verticalItemSpacingInside $leadItemSpacingInside;
        }
        &-icon {
            display: block;
            flex-grow: 0;
            width: 1em;
        }
        &.has-nested-items > &-content > &-icon {
            background-color: #7f7f7f;
            -webkit-clip-path: polygon(25% 25%, 75% 50%, 25% 75%);
            clip-path: polygon(25% 25%, 75% 50%, 25% 75%);
        }
    }
    &-column.trailing-selection &-item.selected {
        background-color: $trailingSelectionColor;
        color: white;
    }
    &-column.trailing-selection &-item.selected &-item-icon:after {
        color: inherit;
    }
    &-details {
        width: $detailsWidth;
        min-width: $detailsWidth;
        overflow: auto;
        border-left: $innerBorder;

        &-header:first-child {
            margin-top: 0;
        }
        &-content {
            padding: 1em;
        }
        &-form {
            display: flex;
            flex-direction: column;
            $labelWidth: 30%;

            &-row {
                display: flex;
                flex-direction: row;
                margin-bottom: 0.5em;
            }
            &-label,
            &-value {
                display: inline-block;
                vertical-align: top;
                word-wrap: break-word;
            }
            &-label {
                width: $labelWidth;
                padding-right: 0.5em;
            }
            &-value {
                width: 100% - $labelWidth;
            }
        }
    }
    &-header {
        flex-direction: row-reverse;
        margin-bottom: $verticalSpacingOutside;
    }
    &-search {
        padding: $spacingInside;

        &-input {
            font: inherit;
            line-height: 2em;   
            border: $innerBorder;
            border-radius: $borderRadius;
            padding: 0 $spacingInside;
            min-width: ($detailsWidth - (2 * $spacingInside));
        }
    }
    &-footer {
        margin-top: $verticalSpacingOutside;
    }
    &-breadcrumbs {
        display: flex;
        flex-flow: row wrap;
        padding: $spacingInside 0.75em $spacingInside (5 * $spacingInside);
        font-family: 'Courier New', Courier, monospace;

        &-icon,
        &-item {
            display: block;
            border-radius: $borderRadius;
        }
        &-icon {
            background-color: $trailingSelectionColor;
            height: 1.5em;
            width: 0.8em;
            margin: 0.25em 0 0.25em (-4 * $spacingInside);
            -webkit-clip-path: polygon(0% 0%, 100% 50%, 0% 100%);
            clip-path: polygon(0% 0%, 100% 50%, 0% 100%);
        }
        &-item {
            background-color: $selectionColor;
            margin-right: -0.75em;
            padding: 0 0.25em 0 1.25em;
            line-height: 2em;
            text-overflow: ellipsis;

            &.has-nested-items {
                padding-right: 1em;
                -webkit-clip-path: polygon(0% 5%, calc(100% - 1em) 5%, 100% 50%, calc(100% - 1em) 95%, 0% 95%, 1em 50%);
                clip-path: polygon(0% 5%, calc(100% - 1em) 5%, 100% 50%, calc(100% - 1em) 95%, 0% 95%, 1em 50%);
            }
            &:not(.has-nested-items) {
                -webkit-clip-path: polygon(0% 5%, 100% 5%, 100% 95%, 0% 95%, 1em 50%);
                clip-path: polygon(0% 5%, 100% 5%, 100% 95%, 0% 95%, 1em 50%);
            }
        }
    }
}
