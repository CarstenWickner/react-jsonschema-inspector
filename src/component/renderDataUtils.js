import PropTypes from "prop-types";

import JsonSchema from "../model/JsonSchema";
import JsonSchemaGroup from "../model/JsonSchemaGroup";
import {
    createGroupFromSchema, getPropertiesFromSchemaGroup, getTypeOfArrayItemsFromSchemaGroup,
    createOptionTargetArrayFromIndexes, getOptionsInSchemaGroup, getIndexPermutationsForOptions
} from "../model/schemaUtils";
import { isDefined, isNonEmptyObject, mapObjectValues } from "../model/utils";
import { createFilterFunctionForSchema } from "../model/searchUtils";

function isOptionIndexValidForOptions(optionIndexes, options) {
    let optionsPart = options;
    optionIndexes.forEach((index) => {
        if (optionsPart && optionsPart.options && optionsPart.options.length > index) {
            optionsPart = optionsPart.options[index];
        } else {
            optionsPart = null;
        }
    });
    return optionsPart && !optionsPart.options;
}

function createRootColumnData(schemas, referenceSchemas, parserConfig) {
    // first prepare those schemas that may be referenced by the displayed ones or each other
    const referenceScopes = [];
    referenceSchemas.forEach((rawRefSchema) => {
        const newReferenceScope = new JsonSchema(rawRefSchema, parserConfig).scope;
        referenceScopes.forEach((otherReferenceScope) => {
            newReferenceScope.addOtherScope(otherReferenceScope);
            otherReferenceScope.addOtherScope(newReferenceScope);
        });
        referenceScopes.push(newReferenceScope);
    });
    // the first column always lists all top-level schemas
    const rootColumnItems = mapObjectValues(schemas, (rawSchema) => {
        const schema = new JsonSchema(rawSchema, parserConfig);
        referenceScopes.forEach(referenceScope => schema.scope.addOtherScope(referenceScope));
        return createGroupFromSchema(schema);
    });
    return { items: rootColumnItems };
}

function buildDefaultArrayProperties(arrayItemSchema) {
    return { "[0]": arrayItemSchema };
}

function buildNextColumn(schemaGroup, optionIndexes, buildArrayProperties = buildDefaultArrayProperties) {
    if (!optionIndexes) {
        const options = getOptionsInSchemaGroup(schemaGroup);
        if (options.options) {
            // next column should offer the selection of options within the schema group
            return {
                contextGroup: schemaGroup,
                options
            };
        }
    }
    // next column should list all available properties
    const propertySchemas = getPropertiesFromSchemaGroup(schemaGroup, optionIndexes);
    if (isNonEmptyObject(propertySchemas)) {
        // convert the individual JsonSchema values into JsonSchemaGroups
        return {
            items: mapObjectValues(propertySchemas, createGroupFromSchema)
        };
    }
    // there are no properties, so this might be an array
    const nestedArrayItemSchema = getTypeOfArrayItemsFromSchemaGroup(schemaGroup, optionIndexes);
    if (nestedArrayItemSchema) {
        // next column should allow accessing the schema of the array's items
        const arrayProperties = mapObjectValues(
            buildArrayProperties(nestedArrayItemSchema),
            propertyValue => (
                propertyValue instanceof JsonSchema ? propertyValue : new JsonSchema(propertyValue, nestedArrayItemSchema.parserConfig)
            )
        );

        return {
            items: mapObjectValues(arrayProperties, createGroupFromSchema)
        };
    }
    return {};
}

export function createRenderDataBuilder(onSelectInColumn) {
    return (schemas, referenceSchemas, selectedItems, parserConfig, buildArrayProperties) => {
        // the first column always lists all top-level schemas
        let nextColumn = createRootColumnData(schemas, referenceSchemas, parserConfig);
        let selectedSchemaGroup;
        const columnData = selectedItems.map((selection, index) => {
            const currentColumn = nextColumn;
            const isOptionSelection = typeof selection !== "string";
            let isValidSelection;
            if (isOptionSelection && selectedSchemaGroup && currentColumn.options) {
                isValidSelection = isOptionIndexValidForOptions(selection, currentColumn.options);
            } else if (!isOptionSelection && currentColumn.items) {
                selectedSchemaGroup = currentColumn.items[selection];
                isValidSelection = isDefined(selectedSchemaGroup);
            }
            if (isValidSelection) {
                nextColumn = buildNextColumn(selectedSchemaGroup, isOptionSelection ? selection : undefined, buildArrayProperties);
                if (isOptionSelection) {
                    selectedSchemaGroup = null;
                }
            } else {
                nextColumn = {};
            }
            // name of the selected item (i.e. key in 'items') or int array of option indexes
            currentColumn.selectedItem = isValidSelection ? selection : null;
            currentColumn.onSelect = onSelectInColumn(index);
            return currentColumn;
        }).filter(({ items, options }) => items || options);
        // set the flag for the last column containing a valid selection
        const columnCount = columnData.length;
        if (columnCount) {
            // there is at least one column, check whether the last column has a valid selection
            const selectedItemInLastColumn = columnData[columnCount - 1].selectedItem;
            // if the last column has no valid selection, the second to last column must have one
            if (selectedItemInLastColumn || columnCount > 1) {
                // there is at least one column with a valid selection, mark the column with the trailing selection as such
                columnData[selectedItemInLastColumn ? (columnCount - 1) : (columnCount - 2)].trailingSelection = true;
            }
        }
        // append last column where there is no selection yet, unless the last selected item has no nested items or options of its own
        if (isNonEmptyObject(nextColumn)) {
            nextColumn.onSelect = onSelectInColumn(selectedItems.length);
            columnData.push(nextColumn);
        }
        // wrap the result into a new object in order to make this more easily extensible in the future
        return { columnData };
    };
}

/**
 * Check whether a given JsonSchema has any nested properties or is an array.
 *
 * @param {JsonSchema} schema single schema instance to check in (ignoring any nested sub schemas, e.g. `allOf`/`oneOf`/`anyOf`)
 * @return {Boolean} return whether any properties are mentioned or `schema` represents an array with items of a certain type
 */
function hasSchemaNestedItems(schema) {
    const { schema: rawSchema } = schema;
    const { properties, required, items, additionalItems } = rawSchema;
    return isNonEmptyObject(properties)
        || required
        || isNonEmptyObject(items)
        || isNonEmptyObject(additionalItems);
}

/**
 * Check whether a given JsonSchemaGroup has any nested properties, is an array or contains selectable optional sub-schemas.
 *
 * @param {JsonSchemaGroup} schemaGroup single schema group instance to check in
 * @return {Boolean} return whether any properties are mentioned, it represents an array with items of a certain type or contains selectable options
 */
export function hasSchemaGroupNestedItems(schemaGroup, optionIndexes) {
    return schemaGroup.someEntry(hasSchemaNestedItems, createOptionTargetArrayFromIndexes(optionIndexes))
        || (!optionIndexes && getOptionsInSchemaGroup(schemaGroup).options);
}

const optionShape = {
    groupTitle: PropTypes.string
};
optionShape.options = PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.shape({}), PropTypes.shape(optionShape)])).isRequired;

/**
 * Proptype validation of the `contextGroup` prop (and the related `options`/`items` props).
 *
 * @param {Object} param0 props
 * @param {?JsonSchemaGroup} param0.contextGroup group to validate
 * @param {?Object} param0.options representation of options within the `contextGroup`
 * @param {?Object} param0.items list of selectable entries being listed (i.e. properties/options/array-item-accessors)
 * @return {Error|null} validation error or null if validation is successful
 */
function contextGroupValidator({ contextGroup, options, items }) {
    if (isDefined(contextGroup) && !(contextGroup instanceof JsonSchemaGroup)) {
        return new Error("`contextGroup` is not a JsonSchemaGroup");
    }
    if (isNonEmptyObject(options) !== isDefined(contextGroup)) {
        return new Error("`options` and `contextGroup` are expected to be both set or not at the same time");
    }
    if (isNonEmptyObject(items) === isNonEmptyObject(options)) {
        return new Error("exactly one of `items` or `options` should be provided");
    }
    return null;
}

/**
 * PropType validation of the `selectedItem` prop.
 *
 * @param {Object} param0 props
 * @param {?String|Array.<Number>} param0.selectedItem name of the selected property/array-item-accessor or array of option indexes
 * @return {Error|null} validation error or null if validation is successful
 */
function selectedItemValidator({ selectedItem }) {
    if (isDefined(selectedItem)
        && typeof selectedItem !== "string"
        && !(
            Array.isArray(selectedItem) && selectedItem.length && selectedItem.every(arrayEntry => typeof arrayEntry === "number")
        )) {
        return new Error("`selectedItem` is not a string or array of numbers");
    }
    // assume all ok
    return null;
}

/**
 * PropType validation of the `trailingSelection` prop (and the related `selectedItem` prop).
 *
 * @param {Object} param0 props
 * @param {?String|Array.<Number>} param0.selectedItem name of the selected property/array-item-accessor or array of option indexes
 * @param {?Boolean} param0.trailingSelection flag indicating whether the `selectedItem` is the last selection (i.e. in the right-most column)
 * @return {Error|null} validation error or null if validation is successful
 */
function trailingSelectionValidator({ selectedItem, trailingSelection }) {
    if (isDefined(trailingSelection) && typeof trailingSelection !== "boolean") {
        return new Error("`trailingSelection` is not a boolean");
    }
    if (trailingSelection && !selectedItem) {
        return new Error("`trailingSelection` is true while there is no `selectedItem`");
    }
    return null;
}

/**
 * PropType validation of the `filteredItems` prop.
 *
 * @param {Object} param0 props
 * @param {?Array.<String>|Array.<Array.<Number>>} param0.filteredItems names of properties/options/array-item-accessors matchign the search filter
 * @param {?Object} param0.options representation of options within the `contextGroup`
 * @param {?Object} param0.items list of selectable entries being listed (i.e. properties/options/array-item-accessors)
 * @return {Error|null} validation error or null if validation is successful
 */
function filteredItemsValidator({ items, options, filteredItems }) {
    if (isDefined(filteredItems)) {
        if (!Array.isArray(filteredItems)) {
            return new Error("`filteredItems` is not an `array`");
        }
        if (items && filteredItems.some(singleItem => !items[singleItem])) {
            return new Error("`filteredItems` are not all part of `items`");
        }
        if (options && !filteredItems.every(singleOption => isOptionIndexValidForOptions(singleOption, options))) {
            return new Error("`filteredItems` are not all part of index combinations derived from `options`");
        }
    }
    return null;
}

/**
 * Shape for PropType validation of a single entries in the standard `columnData` array generated by a created render-data-builder.
 *
 * @param {?Boolean} includeOnSelect flag indicating whether `onSelect` should be treated as required
 * @return {Object} return `columnData` shape for PropType validation
 */
export function getColumnDataPropTypeShape(includeOnSelect = true) {
    return {
        items: PropTypes.objectOf(PropTypes.instanceOf(JsonSchemaGroup)),
        options: PropTypes.shape(optionShape),
        contextGroup: contextGroupValidator,
        selectedItem: selectedItemValidator,
        trailingSelection: trailingSelectionValidator,
        filteredItems: filteredItemsValidator,
        onSelect: includeOnSelect ? PropTypes.func.isRequired : PropTypes.func // func(SyntheticEvent: event, string|array<number>: identifier)
    };
}

/**
 * Create a function that returns the list of `filteredItems` for a given entry in the standard `columnData` array.
 *
 * @param {Function} flatSearchFilter function determining whether a given raw json schema matches some search filter
 * @param {Object} flatSearchFilter.param0 raw json schema
 * @param {Boolean} flatSearchFilter.return indication whether raw json schema is matching search filter
 * @return {Function} return function that returns the list of `filteredItems` for a given entry in the standard `columnData` array
 * @return {Object} return.param0 entry in `columnData` array to check
 * @return {?JsonSchemaGroup} return.param0.contextGroup json schema group containing options to select from
 * @return {?Object} return.param0.options representation of options within the `contextGroup`
 * @return {?Object} return.param0.items list of selectable entries being listed (i.e. properties/options/array-item-accessors)
 * @return {?Array.<String>|Array.<Array.<Number>>} return.return names of properties/options/array-item-accessors matchign the search filter
 */
export function createFilterFunctionForColumn(flatSearchFilter) {
    const containsMatchingItems = createFilterFunctionForSchema(flatSearchFilter);
    return ({ items, options, contextGroup }) => {
        if (isNonEmptyObject(items)) {
            return Object.keys(items).filter(key => items[key].someEntry(containsMatchingItems));
        }
        return getIndexPermutationsForOptions(options)
            .filter(optionIndexes => contextGroup.someEntry(containsMatchingItems, createOptionTargetArrayFromIndexes(optionIndexes)));
    };
}
