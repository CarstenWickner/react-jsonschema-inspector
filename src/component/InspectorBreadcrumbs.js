import PropTypes from "prop-types";
import React from "react";
import classNames from "classnames";

import { getColumnDataPropTypeShape } from "./renderDataUtils";

import createBreadcrumbBuilder from "../model/breadcrumbsUtils";

const InspectorBreadcrumbs = ({ columnData, breadcrumbsOptions }) => {
    const buildBreadcrumb = createBreadcrumbBuilder(breadcrumbsOptions);
    const { preventNavigation } = breadcrumbsOptions;
    return (
        <div className="jsonschema-inspector-breadcrumbs">
            <span className="jsonschema-inspector-breadcrumbs-icon" />
            {columnData.map((column, index) => {
                const breadcrumbText = buildBreadcrumb(column, index);
                if (!breadcrumbText) {
                    // omit empty breadcrumb
                    return null;
                }
                const className = classNames({
                    "jsonschema-inspector-breadcrumbs-item": true,
                    "has-nested-items": index < (columnData.length - 2)
                        || (index === (columnData.length - 2)
                            && (!columnData[columnData.length - 1].options || !columnData[columnData.length - 1].selectedItem))
                });
                const { selectedItem, onSelect } = column;
                return (
                    <span
                        // eslint-disable-next-line react/no-array-index-key
                        key={index}
                        className={className}
                        onDoubleClick={preventNavigation ? undefined : event => onSelect(event, selectedItem)}
                    >
                        {breadcrumbText}
                    </span>
                );
            })}
        </div>
    );
};

InspectorBreadcrumbs.propTypes = {
    columnData: PropTypes.arrayOf(PropTypes.shape(getColumnDataPropTypeShape(true))).isRequired,
    breadcrumbsOptions: PropTypes.shape({
        prefix: PropTypes.string,
        separator: PropTypes.string,
        skipSeparator: PropTypes.func,
        mutateName: PropTypes.func,
        preventNavigation: PropTypes.bool
    }).isRequired
};

export default InspectorBreadcrumbs;
