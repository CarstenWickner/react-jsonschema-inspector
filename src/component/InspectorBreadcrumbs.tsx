import * as PropTypes from "prop-types";
import * as React from "react";
import classNames from "classnames";

import { ColumnDataPropType } from "./renderDataUtils";
import { createBreadcrumbBuilder } from "../model/breadcrumbsUtils";
import { BreadcrumbsOptions, RenderColumn, RenderOptionsColumn } from "../types/Inspector";

interface InspectorBreadcrumbsProps {
    columnData: Array<RenderColumn>;
    breadcrumbsOptions: BreadcrumbsOptions;
}

export class InspectorBreadcrumbs extends React.Component<InspectorBreadcrumbsProps> {
    render(): React.ReactNode {
        const { columnData, breadcrumbsOptions } = this.props;
        const buildBreadcrumb = createBreadcrumbBuilder(breadcrumbsOptions);
        const { preventNavigation, renderItem, renderTrailingContent } = breadcrumbsOptions;
        return (
            <>
                <div className="jsonschema-inspector-breadcrumbs">
                    <span className="jsonschema-inspector-breadcrumbs-icon" />
                    {columnData.map((column, index) => {
                        const breadcrumbText = buildBreadcrumb(column, index);
                        if (!breadcrumbText) {
                            // omit empty breadcrumb
                            return null;
                        }
                        const hasNestedItems =
                            index < columnData.length - 2 ||
                            (index === columnData.length - 2 &&
                                (!(columnData[columnData.length - 1] as RenderOptionsColumn).options ||
                                    !columnData[columnData.length - 1].selectedItem));
                        if (renderItem) {
                            return renderItem({
                                breadcrumbText,
                                hasNestedItems,
                                column,
                                index
                            });
                        }
                        const { selectedItem, onSelect } = column;
                        return (
                            <span
                                key={index}
                                className={classNames({
                                    "jsonschema-inspector-breadcrumbs-item": true,
                                    "has-nested-items": hasNestedItems
                                })}
                                onDoubleClick={preventNavigation ? undefined : (event): void => onSelect(event, selectedItem)}
                            >
                                {breadcrumbText}
                            </span>
                        );
                    })}
                </div>
                {renderTrailingContent &&
                    renderTrailingContent({
                        breadcrumbTexts: columnData.map(buildBreadcrumb).filter((b) => !!b),
                        columnData
                    })}
            </>
        );
    }

    static propTypes = {
        columnData: ColumnDataPropType.isRequired,
        breadcrumbsOptions: PropTypes.shape({
            prefix: PropTypes.string,
            separator: PropTypes.string,
            skipSeparator: PropTypes.func,
            mutateName: PropTypes.func,
            preventNavigation: PropTypes.bool,
            renderItem: PropTypes.func,
            renderTrailingContent: PropTypes.func
        }).isRequired
    };
}
