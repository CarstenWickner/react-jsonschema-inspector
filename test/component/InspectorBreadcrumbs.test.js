import React from "react";
import { shallow } from "enzyme";

import InspectorBreadcrumbs from "../../src/component/InspectorBreadcrumbs";
import { createRenderDataBuilder } from "../../src/component/renderDataUtils";

describe("renders correctly", () => {
    const buildColumnData = createRenderDataBuilder(() => () => { });
    it("with minimal/default props", () => {
        const schema = {
            properties: {
                foo: true,
                bar: true
            }
        };
        const { columnData } = buildColumnData({ foobar: schema }, [], ["foobar"]);
        const component = shallow(
            <InspectorBreadcrumbs
                columnData={columnData}
                breadcrumbsOptions={{}}
            />
        );
        expect(component).toMatchSnapshot();
    });
    it("without selection", () => {
        const { columnData } = buildColumnData({ foo: {} }, [], []);
        const component = shallow(
            <InspectorBreadcrumbs
                columnData={columnData}
                breadcrumbsOptions={{}}
            />
        );
        expect(component.find(".jsonschema-inspector-breadcrumbs-icon").exists()).toBe(true);
        expect(component.find(".jsonschema-inspector-breadcrumbs-item").exists()).toBe(false);
        expect(component.text()).toEqual("");
    });
    it("with prefix", () => {
        const { columnData } = buildColumnData({ foo: {} }, [], ["foo"]);
        const component = shallow(
            <InspectorBreadcrumbs
                columnData={columnData}
                breadcrumbsOptions={{
                    prefix: "variableContext."
                }}
            />
        );
        expect(component.find(".jsonschema-inspector-breadcrumbs-item")).toHaveLength(1);
        expect(component.text()).toEqual("variableContext.foo");
    });
    describe("schema with array", () => {
        const schemas = {
            foo: {
                items: {
                    properties: {
                        bar: { title: "Property Title" }
                    }
                }
            }
        };
        it.each`
            testTitle                 | selectedItems            | itemCount | breadcrumbsText
            ${"root selection"}       | ${["foo"]}               | ${1}      | ${"foo"}
            ${"array selection"}      | ${["foo", "[0]"]}        | ${2}      | ${"foo.[0]"}
            ${"array item selection"} | ${["foo", "[0]", "bar"]} | ${3}      | ${"foo.[0].bar"}
        `("$testTitle", ({ selectedItems, itemCount, breadcrumbsText }) => {
            const { columnData } = buildColumnData(schemas, [], selectedItems);
            const component = shallow(
                <InspectorBreadcrumbs
                    columnData={columnData}
                    breadcrumbsOptions={{}}
                />
            );
            expect(component.text()).toEqual(breadcrumbsText);
            expect(component.find(".jsonschema-inspector-breadcrumbs-item")).toHaveLength(itemCount);
        });
        it("with custom separator", () => {
            const { columnData } = buildColumnData(schemas, [], ["foo", "[0]", "bar"]);
            const component = shallow(
                <InspectorBreadcrumbs
                    columnData={columnData}
                    breadcrumbsOptions={{
                        prefix: "$this->",
                        separator: "->"
                    }}
                />
            );
            expect(component.find(".jsonschema-inspector-breadcrumbs-item")).toHaveLength(3);
            expect(component.text()).toEqual("$this->foo->[0]->bar");
        });
    });
    describe("with option selection", () => {
        const schema = {
            oneOf: [
                { title: "Foobar" },
                {
                    properties: {
                        baz: true
                    }
                }
            ]
        };

        it("is skipping option selection", () => {
            const { columnData } = buildColumnData({ foo: schema }, [], ["foo", [1], "baz"]);
            const component = shallow(
                <InspectorBreadcrumbs
                    columnData={columnData}
                    breadcrumbsOptions={{}}
                />
            );
            expect(component.find(".jsonschema-inspector-breadcrumbs-item")).toHaveLength(2);
            expect(component.text()).toEqual("foo.baz");
        });
        it.each`
            testTitle                                                                      | optionIndexes | hasNestedItems
            ${"indicates on previous breadcrumb when selected option has no nested items"} | ${[0]}        | ${false}
            ${"indicates on previous breadcrumb when selected option has nested property"} | ${[1]}        | ${true}
        `("$testTitle", ({ optionIndexes, hasNestedItems }) => {
            const { columnData } = buildColumnData({ foo: schema }, [], ["foo", optionIndexes]);
            const component = shallow(
                <InspectorBreadcrumbs
                    columnData={columnData}
                    breadcrumbsOptions={{}}
                />
            );
            const singleItem = component.find(".jsonschema-inspector-breadcrumbs-item");
            expect(singleItem).toHaveLength(1);
            expect(singleItem.hasClass("has-nested-items")).toBe(hasNestedItems);
            expect(component.text()).toEqual("foo");
        });
    });
    describe("with custom item render function", () => {
        const { columnData } = buildColumnData(
            {
                foo: {
                    properties: {
                        bar: {}
                    }
                }
            },
            [],
            ["foo", "bar"],
            {}
        );
        const renderItem = (breadcrumbText, hasNestedItems, column, index) => (
            <span key={index} className="custom-breadcrumbs-item">
                {`${index + 1}. ${breadcrumbText} (${columnData[index] === column})${hasNestedItems ? " >" : ""}`}
            </span>
        );
        const component = shallow(
            <InspectorBreadcrumbs
                columnData={columnData}
                breadcrumbsOptions={{ renderItem }}
            />
        );
        expect(component.find(".jsonschema-inspector-breadcrumbs-icon").exists()).toBe(true);
        expect(component.find(".jsonschema-inspector-breadcrumbs-item").exists()).toBe(false);
        const customItems = component.find(".custom-breadcrumbs-item");
        expect(customItems).toHaveLength(2);
        expect(customItems.at(0).text()).toEqual("1. foo (true) >");
        expect(customItems.at(1).text()).toEqual("2. .bar (true)");
    });
    describe("with custom item render function", () => {
        const { columnData } = buildColumnData(
            {
                foo: {
                    properties: {
                        bar: {}
                    }
                }
            },
            [],
            ["foo", "bar"],
            {}
        );
        const renderTrailingContent = (breadcrumbTexts, columnDataParam) => (columnDataParam !== columnData ? null : (
            <button type="button">
                {`Copy to Clipboard: ${breadcrumbTexts.join("")}`}
            </button>
        ));
        const component = shallow(
            <InspectorBreadcrumbs
                columnData={columnData}
                breadcrumbsOptions={{ renderTrailingContent }}
            />
        );
        expect(component.find(".jsonschema-inspector-breadcrumbs-icon").exists()).toBe(true);
        expect(component.find(".jsonschema-inspector-breadcrumbs-item")).toHaveLength(2);
        const trailingButton = component.find("button");
        expect(trailingButton.exists()).toBe(true);
        expect(trailingButton.text()).toEqual("Copy to Clipboard: foo.bar");
    });
});
describe("handles double-click navigation", () => {
    const onSelectOne = jest.fn(() => { });
    const onSelectTwo = jest.fn(() => { });
    const { columnData } = createRenderDataBuilder(
        index => (index === 0 ? onSelectOne : onSelectTwo)
    )(
        {
            foo: {
                properties: {
                    bar: {}
                }
            }
        },
        [],
        ["foo", "bar"],
        {}
    );

    it("triggers onSelect on root selection", () => {
        const component = shallow(
            <InspectorBreadcrumbs
                columnData={columnData}
                breadcrumbsOptions={{}}
            />
        );
        const selectedItems = component.find(".jsonschema-inspector-breadcrumbs-item");
        expect(selectedItems).toHaveLength(2);

        selectedItems.at(0).prop("onDoubleClick")({});
        expect(onSelectOne.mock.calls).toHaveLength(1);
        expect(onSelectOne.mock.calls[0][1]).toEqual("foo");
        expect(onSelectTwo.mock.calls).toHaveLength(0);
    });
    it("triggers onSelect on non-root selection", () => {
        const component = shallow(
            <InspectorBreadcrumbs
                columnData={columnData}
                breadcrumbsOptions={{}}
            />
        );
        const selectedItems = component.find(".jsonschema-inspector-breadcrumbs-item");
        expect(selectedItems).toHaveLength(2);

        selectedItems.at(1).prop("onDoubleClick")({});
        expect(onSelectOne.mock.calls).toHaveLength(0);
        expect(onSelectTwo.mock.calls).toHaveLength(1);
        expect(onSelectTwo.mock.calls[0][1]).toEqual("bar");
    });
    it("ignored when navigation is prevented", () => {
        const component = shallow(
            <InspectorBreadcrumbs
                columnData={columnData}
                breadcrumbsOptions={{
                    preventNavigation: true
                }}
            />
        );
        const selectedItems = component.find(".jsonschema-inspector-breadcrumbs-item");
        expect(selectedItems).toHaveLength(2);
        expect(selectedItems.at(0).prop("onDoubleClick")).not.toBeDefined();
        expect(selectedItems.at(1).prop("onDoubleClick")).not.toBeDefined();
    });
});
