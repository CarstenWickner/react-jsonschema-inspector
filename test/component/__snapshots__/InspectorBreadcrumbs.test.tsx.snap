// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with minimal/default props 1`] = `
<Fragment>
  <div
    className="jsonschema-inspector-breadcrumbs"
  >
    <span
      className="jsonschema-inspector-breadcrumbs-icon"
    />
    <span
      className="jsonschema-inspector-breadcrumbs-item has-nested-items"
      key="0"
      onDoubleClick={[Function]}
    >
      foobar
    </span>
  </div>
</Fragment>
`;
